### PR TITLE
Datastructure Features

### LINKED ISSUE #4: Extend DBCSR get_block() to return sparse block
- Add optional argument to get_block(): dense : bool = False
- False by default to avoid memory allocation in the default case
- Return the requested block in the CSR format

### LINKED ISSUE #12: Testing DSBSPARSE

### LINKED ISSUE #20: DSBSPARSE more arithmetics
- [ ] Infinite contact boundary blocks matmult

### LINKED ISSUE #27: DSBCOO: First implementation

### LINKED ISSUE #28: Correctly resolve slices in getitem

### PR DESCRIPTION
**Pull request by vetschn**
This should close #27 and related issues.

### CODE REVIEW

**Comment by vincent-maillou** [comment]
As we are supporting `to_dense()` I think we should also support `to_sparse()` here are my reasons why:
- For the `Inv` method of the GF solvers. For the gpu support, we may want to densify the entire matrix on the GPU instead of the CPU (And I know that this dense solver is a bit of a debug thing, but I feel like if we keep it, it should be supported and implemented correctly)
- For debug/compatibility purposes: let say someone have developed a bit of code that operate on a sparse system matrix. Then we could plug and test very easily its idea in QuaTrEx
- I feel like it can happen very quickly that someone want to look at what the system matrix or the self energies looks like on a system that cannot  be exported with to_dense().

**Comment by vetschn** [comment]
I managed to mostly handle the higher stack dimensions. However, there is a very weird bug in the distributed transposition somewhere that I haven't been able to pin down yet. Most of the time the `.dtranspose()` works beautifully, but for some seemingly random combinations of `stack_shape` and `comm.size` the data on certain ranks is wrong.

**Comment by vetschn** [comment]
The GPU aware implementation is working as well now, i.e. all the datastructure test cases pass both with cupy and numpy backend. The data corruption problem mentioned above is not fixed yet and i need to add a couple more tests.

**Comment by vetschn** [comment]
Everything should work now, last steps are increasing test coverage.

One thing to note: Currently, as we have no GPU-aware MPI, `.dtranspose()` does not work with the `cupy` backend. Calling it when using `cupy` will raise segmentation faults (!)

File: src/qttools/datastructures/dsbsparse.py [hunk 1]
```diff
@@ -1,244 +1,538 @@
...
+        # )
+
+        self._data = _block_view(self._data, axis=block_axis)
+        # We need to make sure that the block-view is memory-contiguous.
+        # This does nothing if the data is already contiguous.
+        self._data = xp.ascontiguousarray(self._data)
+
+        synchronize_current_stream()
+        comm.Alltoall(MPI.IN_PLACE, self._data)
+
```
Comment (vetschn) [inline src/qttools/datastructures/dsbsparse.py:328]: @vincent-maillou for the `cupy` backend, should we change this by just transferring the data to host for now? We can try to do more complicated stuff, like detecting if MPI is GPU aware later on... 

```suggestion
        synchronize_current_stream()
        comm.Alltoall(MPI.IN_PLACE, get_host(self._data))
```

**Review by vincent-maillou — APPROVED** [review]
It's awesome!

**Pull request merged**